Making constraint helpers publicly available

murilommen opened the pull request:
## Description
Some helper functions were placed under
python/examples/advanced/Constraints_Suite.ipynb.
This commit makes them available through whylogs.core.constraints import

- [x] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

murilommen committed:
Making constraint helpers publicly available

Details: some helper functions were placed under
python/examples/advanced/Constraints_Suite.ipynb.
This commit makes them available through whylogs.core.constraints import

murilommen committed:
fix circular import issue

TheMellyBee reviewed (approved):
LGTM!

jamie256 commented on python/tests/core/constraints/test_helpers.py:
```diff
@@ -0,0 +1,205 @@
+
+
+@pytest.fixture
+def builder(profile_view):
```
Can we add a test against a NaN value?

murilommen replied:
you mean having a profile_view column with only NaNs and running all tests against it? or just insert some NaNs in the column itself

jamie256 replied:
Right, your nan_builder looks good, as long as there is one NaN value in our test input considering we had seen pitfalls around that scenario. Maybe one bad row, or an empty dataset kind of thing is a realistic scenario if the inputs come in batches over time, we can test that the constraints don't fail when there is a NaN value in there.

The other idea was to profile an empty numeric column so that the DistributionMetric contained NaN for the max/min  so we can show that the constraint condition wasn't surprising in the results. I'm expected we want to skip failing when evaluating and empty distribution metric with respect to a `greater_than_number ` constraint condition?

murilommen replied:
for the second idea, we can make that optional, as @FelipeAdachi did for `quantile_between_range`. should that be the case for all  metrics then?

jamie256 replied:
its normal to have empty batches of data in some window of time if we are considering streaming scenarios or segments, and I don't think of those cases where there is no data as violating a range based constraint. Having an optional parameter is fine to control skipping NaN.

jamie256 commented on python/whylogs/core/constraints/helpers.py:
```diff
@@ -0,0 +1,252 @@
+    """
+    constraint = MetricConstraint(
+        name=f"{column_name} greater than number {number}",
+        condition=lambda x: x.min >= number,
```
Suggest defining a function for the conditions over using lambdas for readability and ability to add some checking: e.g. a trace/callstack will be easier to read when the function is named and we can check the input is not NaN for example, and then do the comparison.

murilommen replied:
right!

jamie256 reviewed (commented):
Looks close. Wondering if we can come up with a better name for the file than "helpers.py"? Not sure if any of these are better "presets", "conditions", "factories"?

FelipeAdachi commented on python/whylogs/core/constraints/helpers.py:
```diff
@@ -0,0 +1,252 @@
+from typing import Union
```
But even the currently ones could be distinguished by the metric it uses.

murilommen replied:
firstly i had created different namespaces for each metric, but i ended up having to repeat all the imports in the modules and fixtures to create the very similar tests.

FelipeAdachi replied:
Yes, I don't have a clear suggestion yet.
Maybe separating by each metric is too much granularity?

Right now, some come to mind:
- metric_constraints.py
   - multiple_metrics_constraint (this could live with the metrics_constraints, I guess)
- multiple_columns_constraints.py
- multiple_profile_constraints (some constraints need a baseline, like constraints for ks_test and chi_squared. whether this should be, in fact, a constraint, is also a relevant discussion)
- dataset_constraints.py

But since we have only metric_constraints, we could create a namespace for it and make the other ones as we implement them?

murilommen replied:
yes, and they should all live either on 
whylogs.core.constraints (the actual classes that define them)
or
whylogs.core.constraints.factories (the factories with helper functions to create a constraints suite)

FelipeAdachi replied:
I think they should live under `factories` and leave `core.constraints` for the "building blocks".

murilommen committed:
Refactoring with namespaces

Details: as good suggestions came in the PR reviews, now all helpers will live under
`factories/(metric_namespace).py`

murilommen committed:
fixing imports on example notebook

jamie256 commented on python/whylogs/core/constraints/factories/cardinality_metrics.py:
```diff
@@ -0,0 +1,30 @@
+    """
+
+    def distinct_in_range(metric: Metric):
+        distinct_estimate = metric.hll.value.get_estimate()
```
if the hll metric is empty does this throw?

murilommen replied:
![image](https://user-images.githubusercontent.com/48736656/181637806-32e8138a-8d83-4cfc-ba32-7bd981dd412f.png)
this was from a 10 elements np.float type pandas DataFrame.

FelipeAdachi reviewed (approved):
LGTM!

murilommen committed:
including nan checks on tests

details: reusing code on overall tests, adding NaN check cases and also adding in skip_missing
for when users may wish to pass constraints that don't match the specific metric.
